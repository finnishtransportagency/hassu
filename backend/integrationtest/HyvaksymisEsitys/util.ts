import { config } from "../../src/config";
import { getS3Client } from "../../src/aws/client";
import { DeleteObjectsCommand, ListObjectsV2Command, ListObjectsV2CommandOutput, PutObjectCommand } from "@aws-sdk/client-s3";
import { createPresignedPost } from "@aws-sdk/s3-presigned-post";
import { adaptFileName } from "../../src/tiedostot/paths";
import { getAxios } from "../../src/aws/monitoring";
import FormData from "form-data";

export async function insertYllapitoFileToS3(pathInS3: string) {
  if (pathInS3.includes(config.yllapitoBucketName)) {
    throw new Error("path ei saa sisältää ylläpitoBucketNamea");
  }
  const buffer = Buffer.from("Test");
  await getS3Client().send(
    new PutObjectCommand({
      Body: buffer,
      Bucket: config.yllapitoBucketName,
      Key: pathInS3,
    })
  );
}

export async function insertUploadFileToS3(uuid: string, filename: string) {
  const fileNameWithPath = `${uuid}/${adaptFileName(filename)}`;
  const presignedPost = await createPresignedPost(getS3Client(), {
    Bucket: config.uploadBucketName,
    Key: fileNameWithPath,
    Expires: 600,
  });
  const fields = presignedPost.fields;
  const form = new FormData();
  Object.keys(fields).forEach((key) => {
    form.append(key, fields[key]);
  });
  form.append("file", "Test");
  const axios = getAxios();
  const lengthSync = form.getLengthSync();

  await axios.post(presignedPost.url, form, {
    headers: {
      "Content-Length": lengthSync,
      ...form.getHeaders(),
    },
  });
}

export async function deleteYllapitoFiles(sourcePrefix?: string) {
  if (sourcePrefix && sourcePrefix.includes(config.yllapitoBucketName)) {
    throw new Error("sourcePrefix ei saa sisältää yllapitoBucketNamea");
  }
  const s3 = getS3Client();
  let ContinuationToken = undefined;
  do {
    const { Contents = [], NextContinuationToken }: ListObjectsV2CommandOutput = await s3.send(
      new ListObjectsV2Command({
        Bucket: config.yllapitoBucketName,
        Prefix: sourcePrefix,
        ContinuationToken,
      })
    );
    await deleteFiles(Contents.map(({ Key }) => Key as string));

    ContinuationToken = NextContinuationToken;
  } while (ContinuationToken);
}

export async function emptyUploadFiles() {
  const s3 = getS3Client();
  let ContinuationToken = undefined;
  do {
    const { Contents = [], NextContinuationToken }: ListObjectsV2CommandOutput = await s3.send(
      new ListObjectsV2Command({
        Bucket: config.uploadBucketName,
        ContinuationToken,
      })
    );
    await deleteFiles(
      Contents.map(({ Key }) => Key as string),
      config.uploadBucketName
    );

    ContinuationToken = NextContinuationToken;
  } while (ContinuationToken);
}

async function deleteFiles(paths: string[], bucket: string = config.yllapitoBucketName) {
  if (paths.some((path) => path.includes(config.yllapitoBucketName))) {
    throw new Error("path ei saa sisältää ylläpitoBucketNamea");
  }
  await Promise.all(
    // DeleteObjectsCommand takes at most 1000 items
    getChunksOfThousand(paths).map((paths) =>
      getS3Client().send(
        new DeleteObjectsCommand({
          Bucket: bucket,
          Delete: { Objects: paths.map((path) => ({ Key: path })) },
        })
      )
    )
  );
}

function getChunksOfThousand<T>(array: T[]): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < array.length; i += 1000) {
    chunks.push(array.slice(i, i + 1000));
  }
  return chunks;
}

export async function getYllapitoFilesUnderPath(pathInS3: string): Promise<string[]> {
  const s3 = getS3Client();
  const files: string[] = [];
  let ContinuationToken = undefined;
  do {
    const { Contents = [], NextContinuationToken }: ListObjectsV2CommandOutput = await s3.send(
      new ListObjectsV2Command({
        Bucket: config.yllapitoBucketName,
        Prefix: pathInS3,
        ContinuationToken,
      })
    );
    files.push(...Contents.map(({ Key }) => Key as string));
    ContinuationToken = NextContinuationToken;
  } while (ContinuationToken);
  return files;
}
