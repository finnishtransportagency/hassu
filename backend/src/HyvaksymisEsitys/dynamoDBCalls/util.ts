import { UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { log } from "../../logger";
import { getDynamoDBDocumentClient } from "../../aws/client";
import { ConditionalCheckFailedException } from "@aws-sdk/client-dynamodb";
import { SimultaneousUpdateError } from "hassu-common/error";

export async function sendUpdateCommandToDynamoDB(params: UpdateCommand): Promise<void> {
  if (log.isLevelEnabled("debug")) {
    log.debug("Updating projekti to Hassu with params", { params });
  }
  try {
    const dynamoDBDocumentClient = getDynamoDBDocumentClient();
    await dynamoDBDocumentClient.send(params);
  } catch (e) {
    if (e instanceof ConditionalCheckFailedException) {
      throw new SimultaneousUpdateError("Projektia on p√§ivitetty tietokannassa. Lataa projekti uudelleen.");
    }
    log.error(e instanceof Error ? e.message : String(e), { params });
    throw e;
  }
}
