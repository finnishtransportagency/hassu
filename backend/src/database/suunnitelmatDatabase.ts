import { Suunnitelma } from "../model/suunnitelma";

import { DynamoDB } from "aws-sdk";

const docClient = new DynamoDB.DocumentClient({ apiVersion: "2012-08-10" });

const suunnitelmatTableName = process.env.TABLE_SUUNNITELMAT;

async function createSuunnitelma(suunnitelma: Suunnitelma) {
  const params = {
    TableName: suunnitelmatTableName,
    Item: suunnitelma,
  };
  await docClient.put(params).promise();
}

async function listSuunnitelmat(): Promise<Suunnitelma[]> {
  const params = {
    TableName: suunnitelmatTableName,
  };
  const data = await docClient.scan(params).promise();
  return data.Items as Suunnitelma[];
}

async function getSuunnitelmaById(suunnitelmaId: string): Promise<Suunnitelma> {
  const params = {
    TableName: suunnitelmatTableName,
    Key: { id: suunnitelmaId },
  };
  try {
    const data = await docClient.get(params).promise();
    console.log(data);
    return data.Item as Suunnitelma;
  } catch (e) {
    console.log(e);
    throw e;
  }
}

async function updateSuunnitelma(suunnitelma: Suunnitelma) {
  let updateExpression = "set";
  const ExpressionAttributeNames = {};
  const ExpressionAttributeValues = {};
  for (const property in suunnitelma) {
    if (suunnitelma.hasOwnProperty(property)) {
      if (property === "id") {
        continue;
      }
      updateExpression += ` #${property} = :${property} ,`;
      ExpressionAttributeNames["#" + property] = property;
      ExpressionAttributeValues[":" + property] = suunnitelma[property];
    }
  }

  console.log(ExpressionAttributeNames);

  updateExpression = updateExpression.slice(0, -1);

  const params = {
    TableName: suunnitelmatTableName,
    Key: {
      id: suunnitelma.id,
    },
    UpdateExpression: updateExpression,
    ExpressionAttributeNames,
    ExpressionAttributeValues,
  };

  return docClient
    .update(params)
    .promise()
    .then((result) => {
      return result;
    });
}

export default {
  createSuunnitelma,
  updateSuunnitelma,
  listSuunnitelmat,
  getSuunnitelmaById,
};
