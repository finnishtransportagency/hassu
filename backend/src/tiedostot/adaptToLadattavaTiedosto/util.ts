import { GetObjectCommand } from "@aws-sdk/client-s3";
import { config } from "../../config";
import { getS3Client } from "../../aws/client";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

export function getYllapitoSignedDownloadLink(tiedosto: string): Promise<string> {
  const s3client = getS3Client();
  const getObjectCommand = new GetObjectCommand({
    Bucket: config.yllapitoBucketName,
    Key: tiedosto,
  });
  return getSignedUrl(s3client, getObjectCommand, {
    expiresIn: 60 * 60, // One hour
  });
}
