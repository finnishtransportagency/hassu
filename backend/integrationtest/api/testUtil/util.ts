import { cleanupGeneratedIds } from "./cleanUpFunctions";
import { fileService } from "../../../src/files/fileService";
import {
  AineistoInput,
  AsianTila,
  IlmoitettavaViranomainen,
  Kieli,
  KirjaamoOsoite,
  Status,
  VelhoAineisto,
} from "hassu-common/graphql/apiModel";
import { loadProjektiJulkinenFromDatabase } from "./tests";
import { UserFixture } from "../../../test/fixture/userFixture";
import * as sinon from "sinon";
import { pdfGeneratorClient } from "../../../src/asiakirja/lambda/pdfGeneratorClient";
import { handleEvent as pdfGenerator } from "../../../src/asiakirja/lambda/pdfGeneratorHandler";
import { emailClient } from "../../../src/email/email";
import { Attachment } from "nodemailer/lib/mailer";
import { EnhancedPDF } from "../../../src/asiakirja/asiakirjaTypes";
import { GeneratePDFEvent } from "../../../src/asiakirja/lambda/generatePDFEvent";
import { velho } from "../../../src/velho/velhoClient";
import mocha from "mocha";
import { NotFoundError } from "hassu-common/error";
import { cleanupAnyProjektiData } from "../testFixtureRecorder";
import { expectAwsCalls } from "../../../test/aws/awsMock";
import { CreateScheduleCommand, CreateScheduleCommandInput, ListSchedulesCommand, SchedulerClient } from "@aws-sdk/client-scheduler";
import { handleEvent } from "../../../src/sqsEvents/sqsEventHandlerLambda";
import { Callback, Context } from "aws-lambda";
import { SQSRecord } from "aws-lambda/trigger/sqs";
import assert from "assert";
import SMTPTransport from "nodemailer/lib/smtp-transport";
import { kirjaamoOsoitteetService } from "../../../src/kirjaamoOsoitteet/kirjaamoOsoitteetService";
import {
  openSearchClientIlmoitustauluSyote,
  openSearchClientJulkinen,
  openSearchClientYllapito,
} from "../../../src/projektiSearch/openSearchClient";
import { projektiDatabase } from "../../../src/database/projektiDatabase";
import { assertIsDefined } from "../../../src/util/assertions";
import { lyhytOsoiteDatabase } from "../../../src/database/lyhytOsoiteDatabase";
import crypto from "crypto";
import { EventSqsClientMock } from "./eventSqsClientMock";
import { ProjektiPaths } from "../../../src/files/ProjektiPath";
import fs from "fs";
import { setupLocalDatabase } from "../../util/databaseUtil";
import { personSearchUpdaterClient } from "../../../src/personSearch/personSearchUpdaterClient";
import * as personSearchUpdaterHandler from "../../../src/personSearch/lambda/personSearchUpdaterHandler";
import { velhoCache } from "./cachingVelhoClient";
import { mockClient } from "aws-sdk-client-mock";
import { CloudFront } from "@aws-sdk/client-cloudfront";
import { AloitusKuulutusJulkaisu, KasittelynTila } from "../../../src/database/model";
import MockDate from "mockdate";
import orderBy from "lodash/orderBy";
import { dateTimeToString, nyt, parseDate } from "../../../src/util/dateUtil";
import { parameters } from "../../../src/aws/parameters";
import { mockUUID } from "../../shared/sharedMock";
import { EmailOptions } from "../../../src/email/model/emailOptions";

import { expect } from "chai";
import { ProjektiScheduleManager } from "../../../src/sqsEvents/projektiScheduleManager";
import { asianhallintaService } from "../../../src/asianhallinta/asianhallintaService";
import { SqsEvent } from "../../../src/sqsEvents/sqsEvent";

export async function takeS3Snapshot(oid: string, description: string, path?: string): Promise<void> {
  await takeYllapitoS3Snapshot(oid, description, path);
  await takePublicS3Snapshot(oid, description);
}

export async function takeYllapitoS3Snapshot(oid: string, description: string, path?: string): Promise<void> {
  const obj = await fileService.listYllapitoProjektiFiles(oid, path || "");
  expect({
    ["yllapito S3 files " + description]: cleanupAnyProjektiData(cleanupGeneratedIds(obj)),
  }).toMatchSnapshot(description);
}

export async function takePublicS3Snapshot(oid: string, description: string, path = ""): Promise<void> {
  expect({
    ["public S3 files " + description]: cleanupAnyProjektiData(
      cleanupGeneratedIds(await fileService.listPublicProjektiFiles(oid, path, true))
    ),
  }).toMatchSnapshot(description);
}

export function expectToMatchSnapshot(description: string, obj: unknown): void {
  expect({ description, mockedtime: dateTimeToString(nyt()), obj }).toMatchSnapshot();
}

export function adaptAineistoToInput(aineistot: VelhoAineisto[]): AineistoInput[] {
  return aineistot
    .map((aineisto, index) => {
      const { oid: dokumenttiOid, tiedosto: nimi } = aineisto;
      return { jarjestys: index + 1, nimi, dokumenttiOid };
    })
    .slice(0, 5); // Optimization: don't copy all files
}

export async function expectJulkinenNotFound(oid: string, userFixture: UserFixture): Promise<void> {
  userFixture.logout();
  await expect(loadProjektiJulkinenFromDatabase(oid)).to.eventually.be.rejectedWith(NotFoundError);
  userFixture.loginAs(UserFixture.mattiMeikalainen);
}

export async function expectStatusEiJulkaistu(oid: string, userFixture: UserFixture): Promise<void> {
  userFixture.logout();
  const value = await loadProjektiJulkinenFromDatabase(oid);
  expect(Object.keys(value).length).to.eql(4);
  expect(value.__typename).to.eql("ProjektiJulkinen");
  expect(value.status).to.eql(Status.EI_JULKAISTU);
  expect(value.oid).to.eql(oid);
  expect(Object.keys(value.velho || {}).length).to.eql(1);
  userFixture.loginAs(UserFixture.mattiMeikalainen);
}

export class PDFGeneratorStub {
  private pdfGeneratorLambdaStub!: sinon.SinonStub<[GeneratePDFEvent], Promise<EnhancedPDF>>;
  private pdfs: Pick<EnhancedPDF, "nimi" | "textContent">[] = [];

  constructor() {
    mocha.before(() => {
      this.pdfGeneratorLambdaStub = sinon.stub(pdfGeneratorClient, "generatePDF");
    });
    mocha.beforeEach(() => {
      this.pdfGeneratorLambdaStub.callsFake(async (event) => {
        const pdf: EnhancedPDF = await pdfGenerator(event);
        this.pdfs.push({ nimi: pdf.nimi, textContent: pdf.textContent });
        return pdf;
      });
    });
  }

  verifyAllPDFContents(): void {
    const pdfs = this.pdfs;
    if (pdfs.length > 0) {
      expectToMatchSnapshot(
        "PDF:ien sisällöt",
        pdfs.sort((pdf1, pdf2) => pdf1.nimi.localeCompare(pdf2.nimi))
      );
    }
  }
}

export class EmailClientStub {
  private sendEmailStub!: sinon.SinonStub;
  private sendTurvapostiEmailStub!: sinon.SinonStub;

  constructor() {
    mocha.before(() => {
      this.sendEmailStub = sinon.stub(emailClient, "sendEmail");
      this.sendTurvapostiEmailStub = sinon.stub(emailClient, "sendTurvapostiEmail");
    });
    mocha.beforeEach(() => {
      const fakeEmailSender = (options: EmailOptions) => {
        return Promise.resolve({
          messageId: "messageId_test",
          accepted: (options.to || []) as string[],
          rejected: [],
          pending: [],
        } as unknown as SMTPTransport.SentMessageInfo);
      };
      this.sendEmailStub.callsFake(fakeEmailSender);
      this.sendTurvapostiEmailStub.callsFake(fakeEmailSender);
    });
  }

  verifyEmailsSent(): void {
    function verifyEmailsSentThroughStub(stub: sinon.SinonStub) {
      if (stub.getCalls().length > 0) {
        expect(
          stub
            .getCalls()
            .map((call) => {
              const arg = call.args[0];
              if (arg.attachments) {
                arg.attachments = arg.attachments.map((attachment: Attachment) => {
                  // Remove unnecessary data from snapshot
                  delete attachment.content;
                  delete attachment.contentDisposition;
                  return attachment;
                });
              }
              return arg;
            })
            .sort()
        ).toMatchSnapshot();
        stub.resetHistory();
      }
    }

    verifyEmailsSentThroughStub(this.sendEmailStub);
    verifyEmailsSentThroughStub(this.sendTurvapostiEmailStub);
  }
}

export class CloudFrontStub {
  private stub = mockClient(CloudFront);

  constructor() {
    mocha.before(() => this.stub.reset());
    mocha.beforeEach(() => {
      this.stub.onAnyCommand().resolves({});
    });
    mocha.afterEach(() => this.stub.reset());
  }

  verifyCloudfrontWasInvalidated(expectedNumberOfCalls?: number): void {
    expectAwsCalls("createInvalidation", this.stub.calls(), "CallerReference");
    if (expectedNumberOfCalls) {
      expect(this.stub.calls()).to.have.length(expectedNumberOfCalls);
    }
    this.reset();
  }

  reset(): void {
    this.stub.resetHistory();
  }
}

export class ParametersStub {
  private stub!: sinon.SinonStub;
  asianhallintaEnabled = false;

  constructor() {
    mocha.before(() => {
      this.stub = sinon.stub(parameters, "getParameter");
    });
    mocha.beforeEach(() => {
      this.stub.withArgs("AsianhallintaIntegrationEnabled").callsFake(() => String(this.asianhallintaEnabled));
    });
  }
}

export type SaveProjektiToVelhoMocks = {
  saveKasittelynTilaStub: sinon.SinonStub<[oid: string, kasittelynTila: KasittelynTila], Promise<void>>;
  saveProjektiAloituskuulutusPaivaStub: sinon.SinonStub<[oid: string, aloitusKuulutusJulkaisu: AloitusKuulutusJulkaisu], Promise<void>>;
};

export function mockSaveProjektiToVelho(): SaveProjektiToVelhoMocks {
  const saveKasittelynTilaStub = sinon.stub(velho, "saveKasittelynTila");
  const saveProjektiAloituskuulutusPaivaStub = sinon.stub(velho, "saveProjektiAloituskuulutusPaiva");
  mocha.afterEach(() => {
    if (saveKasittelynTilaStub.getCalls().length > 0) {
      expect({
        "velho.saveKasittelynTila": saveKasittelynTilaStub.getCalls().map((call) => call.args),
      }).toMatchSnapshot();
    }
    if (saveProjektiAloituskuulutusPaivaStub.getCalls().length > 0) {
      expect({
        "velho.saveProjektiAloituskuulutusPaiva": saveProjektiAloituskuulutusPaivaStub.getCalls().map((call) => call.args),
      }).toMatchSnapshot();
    }
    saveProjektiAloituskuulutusPaivaStub.reset();
    saveKasittelynTilaStub.reset();
  });
  return { saveKasittelynTilaStub, saveProjektiAloituskuulutusPaivaStub };
}

export class SchedulerMock {
  private schedulerStub = mockClient(SchedulerClient);
  private schedules: Set<CreateScheduleCommandInput> = new Set<CreateScheduleCommandInput>();

  constructor() {
    mocha.beforeEach(() => {
      this.schedulerStub.on(ListSchedulesCommand).resolves({ Schedules: [] });
    });
    mocha.afterEach(() => {
      this.schedulerStub.reset();
    });
  }

  async verifyAndRunSchedule(): Promise<void> {
    const createCalls = this.schedulerStub.commandCalls(CreateScheduleCommand);
    if (createCalls.length > 0) {
      createCalls
        .map((call) => {
          const input = call.args[0].input as unknown as CreateScheduleCommandInput;
          input.GroupName = "***unittest***";
          return input;
        })
        .forEach((call) => this.schedules.add(call));
    }
    expect(Array.from(this.schedules).sort()).toMatchSnapshot();
    await Array.from(this.schedules).reduce((promiseChain, args: CreateScheduleCommandInput) => {
      return promiseChain.then(async () => {
        assert(args.Target?.Input, "args.Target.Input pitäisi olla olemassa");
        const event: SqsEvent = JSON.parse(args.Target?.Input);
        if (!event.date || (event.date && parseDate(event.date).isBefore(nyt()))) {
          const sqsRecord: SQSRecord = { body: args.Target.Input } as unknown as SQSRecord;
          this.schedules.delete(args);
          await handleEvent({ Records: [sqsRecord] }, undefined as unknown as Context, undefined as unknown as Callback);
          return;
        }
      });
    }, Promise.resolve());
    this.schedulerStub.resetHistory();
  }
}

export function mockKirjaamoOsoitteet(): void {
  let kirjaamoOsoitteetStub: sinon.SinonStub;
  mocha.before(() => {
    kirjaamoOsoitteetStub = sinon.stub(kirjaamoOsoitteetService, "listKirjaamoOsoitteet");
  });
  mocha.beforeEach(() => {
    const osoitteet: KirjaamoOsoite[] = [
      {
        __typename: "KirjaamoOsoite",
        sahkoposti: "uudenmaan_kirjaamo@uudenmaan.ely",
        nimi: IlmoitettavaViranomainen.UUDENMAAN_ELY,
      },
      {
        __typename: "KirjaamoOsoite",
        sahkoposti: "kirjaamo@vayla.vayla",
        nimi: IlmoitettavaViranomainen.VAYLAVIRASTO,
      },
    ];
    kirjaamoOsoitteetStub.resolves(osoitteet);
  });
}

function mockOpenSearch() {
  const queryMocks: sinon.SinonStub[] = [];
  mocha.before(() => {
    queryMocks.push(sinon.stub(openSearchClientYllapito, "query"));
    sinon.stub(openSearchClientYllapito, "deleteDocument");
    sinon.stub(openSearchClientYllapito, "putDocument");
    queryMocks.push(sinon.stub(openSearchClientIlmoitustauluSyote, "query"));
    sinon.stub(openSearchClientIlmoitustauluSyote, "deleteDocument");
    sinon.stub(openSearchClientIlmoitustauluSyote, "putDocument");

    queryMocks.push(sinon.stub(openSearchClientJulkinen[Kieli.SUOMI], "query"));
    sinon.stub(openSearchClientJulkinen[Kieli.SUOMI], "deleteDocument");
    sinon.stub(openSearchClientJulkinen[Kieli.SUOMI], "putDocument");

    queryMocks.push(sinon.stub(openSearchClientJulkinen[Kieli.RUOTSI], "query"));
    sinon.stub(openSearchClientJulkinen[Kieli.RUOTSI], "deleteDocument");
    sinon.stub(openSearchClientJulkinen[Kieli.RUOTSI], "putDocument");
  });

  mocha.beforeEach(() => {
    queryMocks.forEach((qm) => qm.resolves({ status: 200 }));
  });
}

let lyhytOsoiteStub: sinon.SinonStub;

function mockLyhytOsoite() {
  mocha.before(() => {
    lyhytOsoiteStub = sinon.stub(lyhytOsoiteDatabase, "randomizeLyhytOsoite");
  });
  mocha.beforeEach(() => {
    // Generoidaan oid:n perusteella aina vakio lyhytosoite, eikä arvota sitä kuten oikeasti.
    lyhytOsoiteStub.callsFake((oid) => {
      return crypto.createHash("shake256", { outputLength: 2 }).update(oid).digest("hex");
    });
  });
}

let readUsersFromSearchUpdaterLambdaStub: sinon.SinonStub;

export function mockPersonSearchUpdaterClient(): void {
  mocha.before(() => {
    readUsersFromSearchUpdaterLambdaStub = sinon.stub(personSearchUpdaterClient, "readUsersFromSearchUpdaterLambda");
  });
  mocha.beforeEach(() => {
    readUsersFromSearchUpdaterLambdaStub.callsFake(async () => {
      return personSearchUpdaterHandler.handleEvent();
    });
  });
}

let checkAsianhallintaStateLambdaStub: sinon.SinonStub;

export function mockAsianhallintaService(): void {
  mocha.before(() => {
    checkAsianhallintaStateLambdaStub = sinon.stub(asianhallintaService, "checkAsianhallintaState");
  });
  mocha.beforeEach(() => {
    checkAsianhallintaStateLambdaStub.callsFake(() => AsianTila.VALMIS_VIENTIIN);
  });
}

function setupMockDate() {
  mocha.beforeEach(() => {
    MockDate.set("2020-01-01");
  });
  // Vaihdettu afterEach:sta afteriksi, jotta kaikki afterEach-hookit tulisivat edelleen ajettua mockatulla ajalla
  mocha.after(() => {
    MockDate.reset();
  });
}

export function defaultMocks(): {
  schedulerMock: SchedulerMock;
  emailClientStub: EmailClientStub;
  eventSqsClientMock: EventSqsClientMock;
  awsCloudfrontInvalidationStub: CloudFrontStub;
  pdfGeneratorStub: PDFGeneratorStub;
  parametersStub: ParametersStub;
} {
  mockKirjaamoOsoitteet();
  mockOpenSearch();
  setupLocalDatabase();
  const schedulerMock = new SchedulerMock();
  const emailClientStub = new EmailClientStub();
  const eventSqsClientMock = new EventSqsClientMock();
  const awsCloudfrontInvalidationStub = new CloudFrontStub();
  const pdfGeneratorStub = new PDFGeneratorStub();
  const parametersStub = new ParametersStub();
  mockLyhytOsoite();
  mockPersonSearchUpdaterClient();
  mockAsianhallintaService();
  setupMockDate();
  velhoCache();
  mockUUID();
  return { schedulerMock, emailClientStub, eventSqsClientMock, awsCloudfrontInvalidationStub, pdfGeneratorStub, parametersStub };
}

export async function verifyProjektiSchedule(oid: string, description: string): Promise<void> {
  const dbProjekti = await projektiDatabase.loadProjektiByOid(oid);
  assertIsDefined(dbProjekti);
  expectToMatchSnapshot(description + ", schedule", orderBy(new ProjektiScheduleManager(dbProjekti).getSchedule(), "date"));
}

export const PATH_EU_LOGO = __dirname + "/../../../../cypress/fixtures/eu-logo.jpg";
export const PATH_KUNTA_LOGO = __dirname + "/../../../../cypress/fixtures/logo.png";

export async function addLogoFilesToProjekti(oid: string): Promise<void> {
  await fileService.createFileToProjekti({
    oid,
    fileName: "suunnittelusopimus/logo.png",
    path: new ProjektiPaths(oid),
    contentType: "image/png",
    contents: fs.readFileSync(PATH_KUNTA_LOGO),
  });

  await fileService.createFileToProjekti({
    oid,
    fileName: "euLogot/FI/logofi.png",
    path: new ProjektiPaths(oid),
    contentType: "image/png",
    contents: fs.readFileSync(PATH_EU_LOGO),
  });

  await fileService.createFileToProjekti({
    oid,
    fileName: "euLogot/SV/logosv.png",
    path: new ProjektiPaths(oid),
    contentType: "image/png",
    contents: fs.readFileSync(PATH_EU_LOGO),
  });
}
