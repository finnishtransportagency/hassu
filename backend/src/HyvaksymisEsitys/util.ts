import { IllegalArgumentError } from "hassu-common/error";
import { DBProjekti } from "../database/model";
import { projektiDatabase } from "../database/projektiDatabase";
import { requirePermissionLuku } from "../user";
import { NykyinenKayttaja } from "hassu-common/graphql/apiModel";

export async function varmistaLukuoikeusJaHaeProjekti(
  oid: string
): Promise<{ projektiInDB: DBProjekti; nykyinenKayttaja: NykyinenKayttaja }> {
  const nykyinenKayttaja = requirePermissionLuku();
  const projektiInDB = await projektiDatabase.loadProjektiByOid(oid);
  if (!projektiInDB) {
    throw new IllegalArgumentError(`Projektia oid:lla ${oid} ei l√∂ydy`);
  }
  return { projektiInDB, nykyinenKayttaja };
}
