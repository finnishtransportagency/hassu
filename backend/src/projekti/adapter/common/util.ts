import * as API from "../../../../../common/graphql/apiModel";
import { isDateTimeInThePast } from "../../../util/dateUtil";
import { findJulkaisuWithTila, GenericKuulutus } from "../../projektiUtil";
import { SaameKieli, SuomiRuotsiKieli } from "../../../database/model";

/**
 *
 * @param kuulutusJulkaisut
 * @returns HYVAKSYTTY kuulutus whose kuulutusPaiva is in the past or a MIGRATOITU kuulutus
 */
export function findPublishedKuulutusJulkaisu<J extends GenericKuulutus>(kuulutusJulkaisut: J[] | undefined | null): J | undefined {
  const hyvaksyttyKuulutus = findJulkaisuWithTila(kuulutusJulkaisut, API.KuulutusJulkaisuTila.HYVAKSYTTY);
  if (hyvaksyttyKuulutus?.kuulutusPaiva && isDateTimeInThePast(hyvaksyttyKuulutus.kuulutusPaiva, "start-of-day")) {
    return hyvaksyttyKuulutus;
  }
  return findJulkaisuWithTila(kuulutusJulkaisut, API.KuulutusJulkaisuTila.MIGROITU);
}

export function forSuomiRuotsiDo(func: (kieli: SuomiRuotsiKieli) => void): void {
  for (const kieli in SuomiRuotsiKieli) {
    func(kieli as SuomiRuotsiKieli);
  }
}

export function forEverySaameDo(func: (kieli: SaameKieli) => void): void {
  for (const saame in SaameKieli) {
    func(saame as SaameKieli);
  }
}
