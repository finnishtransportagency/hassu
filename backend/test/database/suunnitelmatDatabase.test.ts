import { describe, it } from "mocha";
import * as sinon from "sinon";
import suunnitelmatDatabase from "../../src/database/suunnitelmatDatabase";
import { SuunnitelmaFixture } from "../fixture/suunnitelmaFixture";
import { DynamoDB } from "aws-sdk";
import { toMatchSnapshot } from "../util/expect-mocha-snapshot";

import * as expect from "expect";

expect.extend({ toMatchSnapshot });

describe("apiHandler", () => {
  afterEach(() => {
    sinon.reset();
    sinon.restore();
  });

  describe("updateSuunnitelma", () => {
    let fixture: SuunnitelmaFixture;

    let updateStub: sinon.SinonStub;

    beforeEach(() => {
      updateStub = sinon.stub(DynamoDB.DocumentClient.prototype, "update");
      // listSuunnitelmatStub = sinon.stub(suunnitelmatDatabase, "listSuunnitelmat");
    });

    beforeEach(() => {
      fixture = new SuunnitelmaFixture();
    });

    describe("updateSuunnitelma", () => {
      it("should pass expected parameters to DynamoDB", async function () {
        updateStub.returns({
          promise() {
            return Promise.resolve({});
          },
        });

        await suunnitelmatDatabase.updateSuunnitelma(fixture.suunnitelma1);

        sinon.assert.calledOnce(updateStub);
        const call = updateStub.getCall(0);
        expect(call.firstArg).toMatchSnapshot(this as any);
      });
    });
  });
});
