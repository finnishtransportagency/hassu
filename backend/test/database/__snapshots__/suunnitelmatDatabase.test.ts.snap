// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiHandler updateSuunnitelma updateSuunnitelma should pass expected parameters to DynamoDB 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#name": "name",
    "#status": "status",
  },
  "ExpressionAttributeValues": Object {
    ":name": "Test Suunnitelma 1",
    ":status": "EI_JULKAISTU",
  },
  "Key": Object {
    "id": "1",
  },
  "TableName": undefined,
  "UpdateExpression": "set #name = :name , #status = :status ",
}
`;
